_posts/2023-07-24-native-startup-rss-troubleshooting.adoc

```diff
@@ -0,0 +1,230 @@
+
+For Quarkus 2.13.4.Final, the flamegraph looks like this:
+
+image::rss-after-mmap.svg[startup rss startup flamegraph for Quarkus 2.13.4.Final]
+
+If we focus on the same area, we observe the count is 5, so that would be roughly 5MB of allocated heap chunks for the `main` thread:
+
+image::rss-after-mmap-allocate.png[heap chunk allocation related mmap calls for Quarkus 2.13.4.Final]
+
+The increase in object allocation between these two Quarkus versions at runtime startup,
+combined with the idiosyncrasies of how thread local allocation works in SubstrateVM,
+is what caused this increase in RSS consumption.
+
+Since then, improvements were made to the components that caused the extra object allocation.
+By Quarkus version 2.13.7.Final, the RSS consumption on startup was back to previous levels.
+We are also looking into ways to improve thread local allocation logic in SubstrateVM.
+
+=== `ObjectAllocationInNewTLAB` JFR event
+
+Another improvement since we carried out this investigation can be found in the area of tooling.
+At the time, the only tooling we could really use here was `perf` or `eBPF`, which is only available in Linux.
+Additions to the emitted JFR events for native applications in latest GraalVM Community versions include the `jdk.ObjectAllocationInNewTLAB` event,
+which is the JFR the event that would have highlighted discrepancies in the Quarkus versions compared above.
+Unfortunately, the Quarkus versions tested here are not compatible with this GraalVM version,
+but we can give it a go with the latest Quarkus version to see what comes out.
+
+First, build the Quarkus native with JFR:
+
+[source,bash]
+----
+$ ./mvnw package -DskipTests -Dquarkus.native.monitoring=jfr -Dnative
+----
+
+To obtain `jdk.ObjectAllocationInNewTLAB` events,
+a custom JFC configuration file is required.
+`jfr configure` can be used to generate it:
+
+[source,bash]
+----
+$ $JAVA_HOME/bin/jfr configure jdk.ObjectAllocationInNewTLAB#enabled=true --output newtlab.jfc
+----
+
+Then, start Quarkus native executable with the necessary JFR flags:
+
+[source,bash]
+----
+$ getting-started/target/getting-started-1.0.0-SNAPSHOT-runner -XX:+FlightRecorder -XX:StartFlightRecording=settings=newtlab.jfc,filename=recording.jfr -Xmx64m
+...
+2023-07-21 12:25:33,739 INFO  [io.quarkus] (main) getting-started 1.0.0-SNAPSHOT native (powered by Quarkus 3.2.1.Final) started in 0.019s. Listening on: http://0.0.0.0:8080
+----
+
+After shutting down Quarkus,
+we can use https://adoptium.net/jmc[Adoptium's Eclipse Mission Control],
+or even the `jfr` command itself to visualize the `jdk.ObjectAllocationInNewTLAB` events:
+
+[source,bash]
+----
+$ $JAVA_HOME/bin/jfr print --events ObjectAllocationInNewTLAB --stack-depth 100 recording.jfr
+----
+
+Amongst the output you will find events such as the following,
+which are related to the flamegraph stacktraces that we highlighted above:
+
+[source,bash]
+----
+jdk.ObjectAllocationInNewTLAB {
```

Comment:
In this example we don't see the call to `OSCommittedMemoryProvider.allocate()` though, could you perhaps use an event including this as well, or is this something not shown in the JFR event for some reason?

Reply:
The discrepancy is due where the JFR event is emitted vs when the `mmap` system call happens. The JFR event gets emitted after the `slowPathNew*` method has finished, and this method is the one that calls `OSCommittedMemoryProvider.allocate()` before the event fires.

I think the JFR event is fine as is, and don't think there's a need to add a new event. However, what I can maybe do is focus the user on searching for `slowPathNew` methods, and that way the similarities between the two approaches should be smaller. WDYT?

Reply:
Yes, either explain the difference, or move the focus to `slowPathNew` (explaining why it's important), it might be obvious to people familiar with GC, but I guess it's not for most people :)

Reply:
The new updates contain a detailed explanation of the slow/fast allocation paths.